object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -514,40 +670,51 @@ private static boolean isTryWithResourcesVariable(LocalVariableNode lhs) {
    * the check fails.
    */
   private void checkMustCall(
-      LocalVarWithTree localVarWithTree, CFStore store, String outOfScopeReason) {
-    List<String> mustCallValue = typeFactory.getMustCallValue(localVarWithTree.tree);
+      Set<LocalVarWithTree> localVarWithTreeSet, CFStore store, String outOfScopeReason) {
+
+    List<String> mustCallValue =
+        typeFactory.getMustCallValue(localVarWithTreeSet.iterator().next().tree);
     // optimization: if there are no must-call methods, we do not need to perform the check
     if (mustCallValue.isEmpty()) {
       return;
     }
     AnnotationMirror cmAnno;
 
-    // sometimes the store is null!  this looks like a bug in checker dataflow.
-    // TODO track down and report the root-cause bug
-    CFValue lhsCFValue = store != null ? store.getValue(localVarWithTree.localVar) : null;
-    if (lhsCFValue != null) { // When store contains the lhs
-      cmAnno =
-          lhsCFValue.getAnnotations().stream()
-              .filter(anno -> AnnotationUtils.areSameByClass(anno, CalledMethods.class))
-              .findAny()
-              .orElse(typeFactory.top);
-    } else {
-      cmAnno =
-          typeFactory
-              .getAnnotatedType(localVarWithTree.localVar.getElement())
-              .getAnnotationInHierarchy(typeFactory.top);
-    }
-
-    if (!calledMethodsSatisfyMustCall(mustCallValue, cmAnno)) {
-      if (!reportedMustCallErrors.contains(localVarWithTree)) {
-        reportedMustCallErrors.add(localVarWithTree);
+    Set<LocalVarWithTree> defsCopy = new LinkedHashSet<>(localVarWithTreeSet);
+    Iterator<LocalVarWithTree> iter = defsCopy.iterator();
+    while (iter.hasNext()) {
+      LocalVarWithTree localVarWithTree = iter.next();
+      // sometimes the store is null!  this looks like a bug in checker dataflow.
+      // TODO track down and report the root-cause bug
+      CFValue lhsCFValue = store != null ? store.getValue(localVarWithTree.localVar) : null;
+
+      if (lhsCFValue != null) { // When store contains the lhs
+        cmAnno =
+            lhsCFValue.getAnnotations().stream()
+                .filter(anno -> AnnotationUtils.areSameByClass(anno, CalledMethods.class))
+                .findAny()
+                .orElse(typeFactory.top);
+      } else {
+        cmAnno =
+            typeFactory
+                .getAnnotatedType(localVarWithTree.localVar.getElement())
+                .getAnnotationInHierarchy(typeFactory.top);
+      }
 
-        checker.reportError(
-            localVarWithTree.tree,
-            "required.method.not.called",
-            formatMissingMustCallMethods(mustCallValue),
-            localVarWithTree.localVar.getType().toString(),
-            outOfScopeReason);
+      if (calledMethodsSatisfyMustCall(mustCallValue, cmAnno)) {
+        return;
+      } else if (!iter.hasNext()) {
```

Comment:
I don't understand this `!iter.hasNext()` check.  Do you only want to report an error if _all_ the variables in the set do not satisfy the must-call requirement?  Or are you trying to report the error on the final variable?

Reply:
I want to check the first case and also report the error for the first variable in the localVarWithTreeSet!

Reply:
I agree we want to check that all variables do not satisfy the must-call requirement before reporting an error.  We should restructure this code so that there is a boolean flag `mustCallSatisfied` that starts out `false`.  Then, if `calledMethodsSatisfyMustCall(mustCallValue, cmAnno)` returns true, set `mustCallSatisfied` to `true` and break out of the loop.  Then, after the loop, if `mustCallSatisfied` is `false`, report the error on the first variable in the set.  That will make the logic clearer.

Reply:
I updated this code to use mustCallSatisfied boolean.